File: 134.Gas Station.cpp
class Solution {
public:
    int canCompleteCircuit(vector<int>& gas, vector<int>& cost) {
        int size = gas.size();
        int preTank = 0;
        int tank = 0;
        int start = 0;
        for(int i = 0; i < size; ++i) {
        	tank += gas[i] - cost[i];
        	if(preTank < 0) {
        		start = i;
        		preTank = gas[i] - cost[i];
        	}
        	else {
        		preTank += gas[i] - cost[i];
        	}
        }
        return tank >= 0 ? start : -1;
    }
};

File: go/134.Gas_Station.go
func canCompleteCircuit(gas []int, cost []int) int {
	total, pre, start := 0, 0, 0
	for i := 0; i < len(gas); i++ {
		total += gas[i] - cost[i]
		if pre < 0 {
			start = i
			pre = gas[i] - cost[i]
		} else {
			pre += gas[i] - cost[i]
		}
	}
	if total < 0 {
		return -1
	}
	return start
}